providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java

```diff
@@ -61,34 +65,27 @@ public GrpcConnector(final FlagdOptions options, final Cache cache, Consumer<Pro
         this.stateConsumer = stateConsumer;
     }
 
-
     /**
      * Initialize the gRPC stream.
+     *
      * @throws RuntimeException if the connection cannot be established.
      */
-    public void initialize() throws RuntimeException {
-        try {
-            // try a dummy request
-            this.serviceBlockingStub
-                    .withWaitForReady()
-                    .withDeadlineAfter(this.deadline, TimeUnit.MILLISECONDS)
-                    .resolveBoolean(Schema.ResolveBooleanRequest.newBuilder().setFlagKey("ready?").build());
-        } catch (StatusRuntimeException e) {
-            // only return the exception if we don't meet the deadline
-            if (Status.DEADLINE_EXCEEDED.equals(e.getStatus())) {
-                throw e;
-            }
-        } finally {
-            // try in background to open the event stream
-            this.startEventStream();
-        }
+    public void initialize() {
+        eventObserverThread = new Thread(this::observeEventStream);
+        eventObserverThread.start();
```

Comment:
This reads much simpler, but we aren't blocking until we actually get a good connection, like I mentioned [here](https://github.com/open-feature/java-sdk-contrib/pull/378#discussion_r1281102577).

This means that flagd can't be reached, we'll see a READY from the SDK (because initialize returns immediately) and then after that get an error from the thread. Unless I'm missing something, this could be fixed simply by waiting on the `sync` here.

Reply:
> This means that flagd can't be reached, we'll see a READY from the SDK

This is incorrect. Let me try to explain.

Flagd provider is an `EventProvider` with a default status of `NOT_READY`. WIth the current refactoring, status is fully maintained by event subscriptions, observed concurrently through this logic [^1]. The actual event subscriptions are handled by `EventStreamObserver` [^2].

Regarding initialization returning immediately, given that we have start status of `NOT_READY`, we are sticking to this status till we perform a callback from the event observer. And this happens fast as soon as event observer attempt to connect with flagd. And there are two possibilities that happens at `EventStreamObserver`

- Successful connection - results in a `ProviderReady` sent by flagd as the initial handshake

This makes `EventStreamObserver` to callback with ready status, making provider ready

- Connection failure - results in invoking `onError`

This makes `EventStreamObserver` to callback with`Error` status, making provider status to `ERROR`

Hope this helps to clarify your doubt. But please let me know if I am not seeing something here :) 


[^1]: https://github.com/open-feature/java-sdk-contrib/blob/1af6f4ce52f326bb12482da906957013acade150/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L114-L144
[^2]: https://github.com/open-feature/java-sdk-contrib/blob/main/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java#L19

Reply:
> Regarding initialization returning immediately, given that we have start status of NOT_READY, we are sticking to this status till we perform a callback from the event observer. And this happens fast as soon as event observer attempt to connect with flagd. And there are two possibilities that happens at EventStreamObserver

Regardless of this, when this `initialize` returns successfully, the SDK fires a READY event. This is what I mean when I said "we'll see a READY from the SDK".

Reply:
Addressed this with the latest commit - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/a21363ead7e869ab9550d180c385dcc36744ea0b

Now we perform a busy wait for a successful connection. This is done with a countdown similar to what you proposed - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/a21363ead7e869ab9550d180c385dcc36744ea0b#diff-67d657d2b09b8caa955560fe7640113ec2bb5528db86e0f96dc15b83e86c9627R172-R182

please check and let me know your thoughts :)

Reply:
looks good!